docs/lsp.md

```diff
@@ -0,0 +1,480 @@
+# F# LSP
+
+F# LSP support design proposal. To be expanded as we learn more / settle on things.
+
+## Table of contents
+
+- [F# LSP](#f-lsp)
+  - [Table of contents](#table-of-contents)
+  - [Goals / Requirements](#goals--requirements)
+  - [Overview](#overview)
+  - [Components](#components)
+    - [FsLSP](#fslsp)
+      - [FsLSPServer](#fslspserver)
+      - [Workspace State Management](#workspace-state-management)
+        - [Things to keep track of:](#things-to-keep-track-of)
+      - [Project model](#project-model)
+      - [Example sequence diagram](#example-sequence-diagram)
+      - [Concurrency and cancellation](#concurrency-and-cancellation)
+      - [Request tracking](#request-tracking)
+    - [NewVsix](#newvsix)
+    - [FCS](#fcs)
+  - [LSP library](#lsp-library)
+    - [Candidates](#candidates)
+  - [LSP client (?)](#lsp-client-)
+  - [Diagnostic mode](#diagnostic-mode)
+  - [LSIF](#lsif)
+  - [Performance](#performance)
+  - [Analyzers \& Refactoring support](#analyzers--refactoring-support)
+  - [Mixed F# / C# solutions](#mixed-f--c-solutions)
+  - [Action plan](#action-plan)
+    - [Phase 0](#phase-0)
+    - [Phase 1](#phase-1)
+    - [Phase 2 - 5](#phase-2---5)
+  - [LSP Endpoints](#lsp-endpoints)
+
+
+## Goals / Requirements
+
+- A single LSP server implementation that can be used by both VS and VS Code (or any other consumer)
```

Comment:
Can this point be clarified a bit? Right now it's unclear if the deliverable is a binary (i.e. that would supplant FSAC) or a library (that FSAC could consume to off-load parts of the protocol but still extend for things that we need).

Reply:
As I mentioned below, I think we can (and will do both), but we don't want to force ionide to use our server, but we would very much like to reuse majority of editor tooling (intelligence, diagnostics, inlay hints, quick fixes, refactoring, custom analyser messages) everywhere, so we don't duplicate efforts in supporting new features.

But I'll leave it to @0101 to clarify.

Reply:
Yeah I think we should provide both, I'll try to clarify it in the document. 

I'm envisioning that the library will contain a fully functional F# LSP server implementation that you can just connect to some in/out streams and it will work. And it's easy to also ship a binary that will just wrap that in an exe so it can be used directly.

But we should probably enable some extensibility so that you can add more endpoints or override existing ones to suit the needs of a particular editor. 

In an ideal world that shouldn't be needed much, as most functionality should be reusable by everyone and should just go directly into the library.

Reply:
I added a section on extensibility in the doc.